library/init/data/array/basic.lean

```diff
@@ -165,8 +165,35 @@ d_array.iterate a b f
 def foreach (a : array n α) (f : fin n → α → α) : array n α :=
 d_array.foreach a f
 
+/-- Auxilliary function for monadically mapping a function over an array. -/
 @[inline]
-def map (f : α → α) (a : array n α) : array n α :=
+def mmap_core {β : Type v} {m : Type v → Type w} [monad m] (a : array n α) (f : α → m β) :
+  ∀ i ≤ n, m (array i β)
+| 0 _ := pure d_array.nil
+| (i+1) h := do
+    bs ← mmap_core i (le_of_lt h),
+    b ← f (a.read ⟨i, h⟩),
+    pure $ bs.push_back b
+
+/-- Monadically map a function over the array. -/
+@[inline]
+def mmap {β : Type v} {m} [monad m] (a : array n α) (f : α → m β) : m (array n β) :=
+a.mmap_core f _ (le_refl _)
+
+/-- Auxilliary function for mapping a function over an array. -/
+@[inline]
+def map_core {β : Type v} (a : array n α) (f : α → m β) :
+  ∀ i ≤ n, m (array i β)
+| 0 _ := d_array.nil
+| (i+1) h := (mmap_core i (le_of_lt h)).push_back $ f (a.read ⟨i, h⟩)
+
+/-- Map a function over the array. -/
+@[inline]
+def map {β : Type v} (a : array n α) (f : α → β) : array n β :=
+a.map_core f _ (le_refl _)
```

Comment:
`map` needs to be implemented in terms of `foreach` because `foreach` has a fast VM implementation.

This is a list of all functions that have a VM implementation: https://github.com/leanprover-community/lean/blob/01063857bb6814374156433e8cbc0c94a9483f52/src/library/vm/vm_array.cpp#L169-L178